File: LogAndroid/TrustarApp/src/main/kotlin/jp/co/toukei/log/trustar/repo/chat/ChatRepository.kt
package jp.co.toukei.log.trustar.repo.chat

import io.reactivex.rxjava3.core.Completable
import io.reactivex.rxjava3.core.Single
import io.reactivex.rxjava3.functions.Consumer
import jp.co.toukei.log.lib.inTransaction
import jp.co.toukei.log.lib.observeOnIO
import jp.co.toukei.log.lib.rxConsumer
import jp.co.toukei.log.lib.subscribeOnIO
import jp.co.toukei.log.trustar.Current
import jp.co.toukei.log.trustar.db.chat.ChatDB
import jp.co.toukei.log.trustar.db.chat.entity.ChatMessage
import jp.co.toukei.log.trustar.db.chat.entity.ChatMessagePending
import jp.co.toukei.log.trustar.rest.model.MessageItem
import jp.co.toukei.log.trustar.rest.post.DefaultPostContent
import jp.co.toukei.log.trustar.rest.post.MessageItemRows
import jp.co.toukei.log.trustar.toCompletable
import java.util.LinkedList

class ChatRepository(@JvmField val chatDB: ChatDB) {

    private val api = Current.chatApi

    fun updateRead(messageRow: String): Completable {
        return api.updateRead(MessageItemRows(listOf(messageRow)))
            .subscribeOnIO()
            .toCompletable()
    }

    fun loadBottomMessage(roomId: String, pageSize: Int): Single<List<MessageItem>> {
        return Single
            .defer {
                val oldest = localLatestSentMessage(roomId, 1).firstOrNull()?.id
                if (oldest != null) api.getMessageNext(roomId, oldest, pageSize)
                else api.getMessageItems(roomId, pageSize)
            }
            .subscribeOnIO()
            .observeOnIO()
            .doOnSuccess(insertMessagesConsumer())
    }

    fun loadTopMessage(roomId: String, pageSize: Int): Single<List<MessageItem>> {
        return Single
            .defer {
                val oldest = chatDB.chatMessageDao().messageOldest(roomId, 1).firstOrNull()?.id
                if (oldest != null) api.getMessagePrevious(roomId, oldest, pageSize)
                else api.getMessageItems(roomId, pageSize)
            }
            .subscribeOnIO()
            .observeOnIO()
            .doOnSuccess(insertMessagesConsumer())
    }

    private fun insertMessagesConsumer(): Consumer<List<MessageItem>> {
        return rxConsumer { insertMessages(it.map(MessageItem::message)) }
    }

    private fun insertMessages(iterable: Iterable<ChatMessage>) {
        val gs = iterable.groupBy { it.roomId }
        val deleted = LinkedList<ChatMessagePending>()
        chatDB.inTransaction {
            gs.forEach { (t, u) ->
                chatMessageDao().insertOrReplace(u)
                deleted += chatMessagePendingDao().selectSent(t)
                chatMessagePendingDao().deleteSent(t)
            }
        }
        deleted.forEach {
            runCatching {
                val att = ChatMessagePending.attachmentExt(it)
                if (att?.isCacheFile == false) {
                    att.file.delete()
                }
            }
        }
    }

    fun deletePendingMessage(pending: ChatMessagePending) {
        deletePendingMessage(listOf(pending))
    }

    fun deleteMessage(msg: ChatMessage): Completable {
        return api.deleteMessage(DefaultPostContent.messageId(msg.id))
            .subscribeOnIO()
            .toCompletable()
            .doOnComplete {
                ChatMessage.attachmentFile(msg)?.file?.delete()
            }
    }

    private fun deletePendingMessage(pending: List<ChatMessagePending>) {
        chatDB.inTransaction {
            chatMessagePendingDao().delete(pending)
        }
        pending.forEach {
            runCatching { ChatMessagePending.attachmentExt(it)?.file?.delete() }
        }
    }

    fun localPendingMessages(roomId: String): List<ChatMessagePending> {
        return chatDB.chatMessagePendingDao().messageByRoom(roomId)
    }

    fun localPendingMessagesPrev(
        roomId: String,
        createdDate: Long,
        includeSelf: Boolean,
    ): List<ChatMessagePending> {
        return if (includeSelf) {
            chatDB.chatMessagePendingDao().messagePrevIncluded(roomId, createdDate)
        } else {
            chatDB.chatMessagePendingDao().messagePrev(roomId, createdDate)
        }
    }

    fun localPendingMessagesNext(roomId: String, createdDate: Long): List<ChatMessagePending> {
        return chatDB.chatMessagePendingDao().messageNext(roomId, createdDate)
    }

    fun localLatestSentMessage(roomId: String, pageSize: Int): List<ChatMessage> {
        return chatDB.chatMessageDao().messageLatest(roomId, pageSize)
    }

    fun localMessagesPrev(roomId: String, pageSize: Int, startR: String): List<ChatMessage> {
        return chatDB.chatMessageDao().messagePrev(roomId, pageSize, startR)
    }

    fun localMessagesNext(roomId: String, pageSize: Int, startR: String): List<ChatMessage> {
        return chatDB.chatMessageDao().messageNext(roomId, pageSize, startR)
    }

    fun localMessagesBetween(roomId: String, startR: String, endR: String): List<ChatMessage> {
        return chatDB.chatMessageDao().messageBetween(roomId, startR, endR)
    }

    fun localMessagesBetweenAround(
        roomId: String,
        startR: String,
        endR: String,
        prepend: Int,
        append: Int,
    ): List<ChatMessage> {
        val p = localMessagesPrev(roomId, prepend, startR)
        val b = localMessagesBetween(roomId, startR, endR)
        val n = localMessagesNext(roomId, append, endR)
        return ArrayList<ChatMessage>(p.size + b.size + n.size).apply {
            addAll(p)
            addAll(b)
            addAll(n)
        }
    }
}

File: LogiOS/TrustarApp/Repo/Chat/ChatRepository.swift
//
//  ChatRepository.swift
//  TrustarApp
//
//  Created by DionSoftware on 2022/11/25.
//  Copyright © 2022 DionSoftware. All rights reserved.
//

import Foundation
import Combine

class ChatRepository {
    private var api: ChatApi
    private var chatDB: ChatDb
    init(api: ChatApi, chatDB: ChatDb) {
        self.api = api
        self.chatDB = chatDB
    }
    
    func updateRead(messageRow: String) -> AnyPublisher<StateResult, NetworkError> {
        return api.updateRead(body: MessageItemRows(messageRowIds: [messageRow], userInfo: Current.Shared.userInfo()), apiKey: Current.Shared.loggedUser?.token ?? "")
    }
    
    func loadBottomMessage(roomId: String, pageSize: Int) -> AnyPublisher<Event<[MessageItem]>, NetworkError> {
        let oldest = localLatestSentMessage(roomId: roomId, pageSize: 1).first?.id
        let task: AnyPublisher<[MessageItem], NetworkError>
        if oldest != nil {
            task = api.getMessageNext(roomId: roomId, messageId: oldest!, pageSize: pageSize, apiKey: Current.Shared.loggedUser?.token ?? "")
        } else {
            task = api.getMessageItems(roomId: roomId, pageSize: pageSize, apiKey: Current.Shared.loggedUser?.token ?? "")
        }
        
        return insertMessagesConsumer(task: task)
    }
    
    func loadTopMessage(roomId: String, pageSize: Int)-> AnyPublisher<Event<[MessageItem]>, NetworkError> {
        let oldest = chatDB.chatMessageDao?.messageOldest(roomId: roomId, size: 1).first?.id
        let task: AnyPublisher<[MessageItem], NetworkError>
        if oldest != nil {
            task = api.getMessagePrevious(roomId: roomId, messageId: oldest!, pageSize: pageSize, apiKey: Current.Shared.loggedUser?.token ?? "")
        } else {
            task = api.getMessageItems(roomId: roomId, pageSize: pageSize, apiKey: Current.Shared.loggedUser?.token ?? "")
        }
        
        return insertMessagesConsumer(task: task)
    }

    private func insertMessagesConsumer(task: AnyPublisher<[MessageItem], NetworkError>) -> AnyPublisher<Event<[MessageItem]>, NetworkError> {
        return task
            .eraseToAnyPublisher()
            .map({ [self] msgs -> AnyPublisher<Event<[MessageItem]>, NetworkError> in
                insertMessages(iterable: msgs.map({ $0.message() }))
                return Just(Event.Value(value: msgs)).setFailureType(to: NetworkError.self)
                    .eraseToAnyPublisher()
            })
            .switchToLatest()
            .replaceError(with: .Error(error: .unknow))
            .setFailureType(to: NetworkError.self)
            .eraseToAnyPublisher()
    }

    private func insertMessages(iterable: [ChatMessage]) {
        let gs = Dictionary.init(grouping: iterable, by: { $0.roomId })
        var deleted: [ChatMessagePending] = []
        
        do {
            for item in gs {
                let t = item.key
                let u = item.value
                try chatDB.instanceDb?.write { db in for c in u { try c.save(db) }}
                deleted += chatDB.chatMessagePendingDao!.selectSent(roomId: t)
                chatDB.chatMessagePendingDao?.deleteSent(roomId: t)
            }
            
            for d in deleted {
                let att = d.attachmentExt(message: d)
                if att?.isCacheFile == false {
                    if FileManager.default.fileExists(atPath: att!.file.path) {
                        do {
                            try FileManager.default.removeItem(atPath: att!.file.path)
                        } catch let removeError {
                            print("couldn't remove file at path", removeError)
                        }
                    }
                }
            }
        } catch {
            debugPrint(error)
        }
    }
    
    func deletePendingMessage(pending: ChatMessagePending) {
        deletePendingMessage(pending: [pending])
    }
    
    func deleteMessage(msg: ChatMessage) -> AnyPublisher<StateResult, NetworkError> {
        let task = api.deleteMessage(body: DefaultPostContent.Message(messageId: msg.id, userInfo: Current.Shared.userInfo()), apiKey: Current.Shared.loggedUser?.token ?? "")
        
        return Future<StateResult, NetworkError> { promise in
            _ = task
                .subscribe(on: DispatchQueue.global())
                .sink(receiveCompletion: { completion in
                    
                }, receiveValue: { rs in
                    
                    if let path = msg.attachmentFile(message: msg)?.file.path {
                        if FileManager.default.fileExists(atPath: path) {
                            try? FileManager.default.removeItem(atPath: path)
                        }
                    }
                    
                    promise(.success(rs))
                })
        }.map {$0}.eraseToAnyPublisher()
    }
    
    private func deletePendingMessage(pending: [ChatMessagePending]) {
        do {
            try chatDB.instanceDb?.write { db in
                for p in pending {
                    try p.delete(db)
                    
                    if let path = p.attachmentExt(message: p)?.file.path {
                        if FileManager.default.fileExists(atPath: path) {
                            try FileManager.default.removeItem(atPath: path)
                        }
                    }
                }
            }
            
        } catch {
            debugPrint(error)
        }
    }
    
    func localPendingMessages(roomId: String) -> [ChatMessagePending] {
        return chatDB.chatMessagePendingDao!.messageByRoom(roomId: roomId)
    }
    
    func localPendingMessagesPrev(
        roomId: String,
        createdDate: Int64,
        includeSelf: Bool
    ) -> [ChatMessagePending] {
        return includeSelf
        ? chatDB.chatMessagePendingDao!.messagePrevIncluded(roomId: roomId, createdDate: createdDate)
        : chatDB.chatMessagePendingDao!.messagePrev(roomId: roomId, createdDate: createdDate)
    }
    
    func localPendingMessagesNext(roomId: String, createdDate: Int64) -> [ChatMessagePending] {
        return chatDB.chatMessagePendingDao!.messageNext(roomId: roomId, createdDate: createdDate)
    }
    
    func localLatestSentMessage(roomId: String, pageSize: Int) -> [ChatMessage] {
        return chatDB.chatMessageDao!.messageLatest(roomId: roomId, size: pageSize)
    }
    
    func localMessagesPrev(roomId: String, pageSize: Int, startR: String) -> [ChatMessage] {
        return chatDB.chatMessageDao!.messagePrev(roomId: roomId, size: pageSize, messageRow: startR)
    }
    
    func localMessagesNext(roomId: String, pageSize: Int, startR: String) -> [ChatMessage] {
        return chatDB.chatMessageDao!.messageNext(roomId: roomId, size: pageSize, messageRow: startR)
    }
    
    func localMessagesBetween(roomId: String, startR: String, endR: String) -> [ChatMessage] {
        return chatDB.chatMessageDao!.messageBetween(roomId: roomId, startR: startR, endR: endR)
    }
    
    func localMessagesBetweenAround(
        roomId: String,
        startR: String,
        endR: String,
        prepend: Int,
        append: Int
    ) -> [ChatMessage] {
//        let p = localMessagesPrev(roomId: roomId, pageSize: prepend, startR: startR)
        let b = localMessagesBetween(roomId: roomId, startR: startR, endR: endR)
        let n = localMessagesNext(roomId: roomId, pageSize: append, startR: endR)
        
        return b + n
    }
}
